### ISSUE TITLE
Should 0 times a Quantity keep the units that the Quantity had? 

### ISSUE THREAD

**Issue by normalhuman** [sympy/sympy#14374]
The current behavior is that multiplying anything (except an infinite quantity) by 0 returns 0 itself. ([Relevant code](https://github.com/sympy/sympy/blob/master/sympy/core/mul.py#L596)).
```
>>> from sympy.physics.units import meter
>>> 0*meter
0
```
Should this be changed so that the result is 0 meters, instead of unitless 0? This was asked on Stack Overflow ([Why doesn't sympy support units with 0 magnitude?
](https://stackoverflow.com/q/49084526)) and seems to make sense to me.

**Comment by Upabjojr** [comment]
Well, you can still have `Mul(0, meter, evaluate=False)`.

Compare this behaviour with complex numbers. We don't distinguish between real zero and complex zero. The imaginary unit is a variable you put in the expression to have complex numbers.

Quantities are currently treated as variables, so a concept analogous to complex numbers apply.

Also consider that the `Quantity` object supports a scale factor definition, which can be set to zero.

**Comment by asmeurer** [comment]
It makes sense mathematically, but I don't know how hard it is to do in implementation. 

> Compare this behaviour with complex numbers. We don't distinguish between real zero and complex zero. The imaginary unit is a variable you put in the expression to have complex numbers.

The analogy isn't apt here because you can add real and imaginary numbers, but you can't add incompatible quantities, even if they have zero magnitude.

**Comment by bjodah** [comment]
FWIW, I would prefer if e.g.
```
>>> 0*meter + 0*seconds
```
raised an Exception.

**Comment by schymans** [comment]
+1 for `0*meter + 0*secons` raising an exception. See also Issue #13267.

**Comment by czgdp1807** [comment]
This issue appears quite interesting. Can I work on this? I am starting my work on this. :-)

**Comment by Upabjojr** [comment]
@czgdp1807 feel free to work on it if you like.

**Comment by Upabjojr** [comment]
Just beware that `meter + second` may be meaningful in relativity.

**Comment by czgdp1807** [comment]
@Upabjojr Sure. That's an important point.

**Comment by schymans** [comment]
@Upabjojr, could you point us to an example? @czgdp1807 see also: https://github.com/sympy/sympy/issues/14318

**Comment by Upabjojr** [comment]
@schymans my wish is to make the units module independent of SI. That is, allow the choice of unit systems. If the user decides to use natural units, `meter+second` is a meaningful expression as one second = 299792458 meters in that unit system.

**Comment by schymans** [comment]
In a unit system where second = 299792458 meters, length and time are not independent dimensions. We should probably only do dimensional consistency checks for independent dimensions. I agree with you that unit checking should not be hard-wired, but related to the unit system. How could this be done?

**Comment by Upabjojr** [comment]
> How could this be done?

That's the hard part. I believe that something similar to the new-style assumptions could be the way to go.

**Comment by schymans** [comment]
How about not raising any exceptions and moving dimensional checking into a separate function, which could be defined in each dimensional system individually? See also: https://github.com/sympy/sympy/pull/14923#pullrequestreview-139982093

**Comment by meganly** [comment]
What about doing something similar to Pint and having a class called `PhysicalQuantity` or `MagnitudeQuantity` to represent numerical quantities with units as `PhysicalQuantity(magnitude, units)`?  Then `PQ(0, meter)` would guard against loss of units when the magnitude is 0.   Multiplication could be defined by doing something like `PQ(3, N) * PQ(2, m) = PQ(6, N*m)`.

**Comment by schymans** [comment]
@meganly something similar to the `PhysicalQuantity` idea has also been proposed here: https://github.com/sympy/sympy/issues/19313
It is currently available as a separate package based on sympy, in case you are interested: https://essm.readthedocs.io

**Comment by smichr** [comment]
> Multiplication could be defined by doing something like `PQ(3, N) * PQ(2, m) = PQ(6, N*m)`.

I think I mentioned that this can all be done transparent to the user via the printer which would print the PQ without the PQ except when the magnitude is 0, so `PQ(6, m)` would show as `6*m`.

**Comment by schymans** [comment]
@smichr could you give a simple example?

**Comment by Upabjojr** [comment]
> What about doing something similar to Pint and having a class called `PhysicalQuantity` or `MagnitudeQuantity` to represent numerical quantities with units as `PhysicalQuantity(magnitude, units)`? Then `PQ(0, meter)` would guard against loss of units when the magnitude is 0. Multiplication could be defined by doing something like `PQ(3, N) * PQ(2, m) = PQ(6, N*m)`.

Yes, that could be an idea. But I would make it slightly different, I would put the dimension as the second parameter. Maybe a class like this:

```python
class QuantityExpr(Expr):

    def __new__(cls, expr, dimension):
        assert isinstance(dimension, Dimension)
        obj = Expr.__new__(cls, expr, dimension)
        return obj

    def __add__(self, other):
        assert isinstance(other, (Quantity, QuantityExpr))
        assert self.dimension == other.dimension
        return QuantityExpr(self.expr + other.expr, self.dimension)

    ... other methods ...
```

In alternative, the `.kind` property has been introduced into SymPy to avoid the need of creating new classes and override the arithmetic operator methods like `__add__`. It could be checked how it works to see if it can be used.

> > Multiplication could be defined by doing something like `PQ(3, N) * PQ(2, m) = PQ(6, N*m)`.
> 
> I think I mentioned that this can all be done transparent to the user via the printer which would print the PQ without the PQ except when the magnitude is 0, so `PQ(6, m)` would show as `6*m`.

Yes, have a look at `UnevaluatedExpr` class... it does something similar in principle.

**Comment by smichr** [comment]
> @smichr could you give a simple example?

In the str.py file would appear something like this:
```python
def _print_PQ(self, expr):
    if expr.magnitude:
        return (%s*%s" % (expr.magnitude, expr.units))
    return "PQ(%s, %s)" % (expr.magnitude, expr.units)
```

**Comment by nopria** [comment]
> The current behavior is that multiplying anything (except an infinite quantity) by 0 returns 0 itself. ([Relevant code](https://github.com/sympy/sympy/blob/master/sympy/core/mul.py#L596)).
> 
> ```
> >>> from sympy.physics.units import meter
> >>> 0*meter
> 0
> ```
> 
> Should this be changed so that the result is 0 meters, instead of unitless 0? This was asked on Stack Overflow ([Why doesn't sympy support units with 0 magnitude?](https://stackoverflow.com/q/49084526)) and seems to make sense to me.

`0*meter` should definitively return `0*meter`, returning a unitless zero is not consistent and might cause troubles when coding, like I'm experiencing.

**Comment by Upabjojr** [comment]
> 0*meter should definitively return 0*meter, returning a unitless zero is not consistent and might cause troubles when coding, like I'm experiencing.

> This issue prevents in practice the use of sympy with physical quantities for almost anything more than playing, I believe it's urgent to fix it.

Are you suggesting to leave `Mul(0, meter)` unevaluated? Maybe this is the quickest fix... but is it really the optimal fix?

**Comment by schniepp** [comment]
> `0*meter` should definitively return `0*meter`, returning a unitless zero is not consistent and might cause troubles when coding, like I'm experiencing.
> 
>     * For example, it is not possible to write a function to determine if a value has dimensions or not, because it could give no answer for value zero (`sympy.core.numbers.Zero`).
> 
>     * Or, another example, you cannot check if all the values in a list have the same unit, because if one or more items in list are zero you cannot tell its unit.
> 
>     * Or if you have a list of homogeneous quantities and you want to convert them to another unit, an item with magnitude zero raise an error because it cannot be converted having no unit.
> 
> 
> These and many more troubles show up immediatly when coding and they all derive from the inconsistency due to removal of unit in the product with zero.
> 
> This issue prevents in practice the use of sympy with physical quantities for almost anything more than playing, I believe it's urgent to fix it.

For what it's worth, I think the second and third items on your list have other solutions:
- A function to determine whether all values in a list have the same unit: a zero would be compatible with any unit, and thus would return "true" in unit comparisons with any quantity.
- If zeros are treated as compatible with any unit, this means that the unit conversion function would simply leave the zero as a zero.

The first one is more tricky: to determine whether a value has a dimension or not. Perhaps, this is not the right question to ask? Where is this needed? Perhaps, for this one, a pseudo-unit particularly made for zeros could be made? If a function receives this unit as a response, it would have to be treated by the code separately. So the quantity zero technically has a unit, but one that is compatible with any other unit.

**Comment by nopria** [comment]
> > 0_meter should definitively return 0_meter, returning a unitless zero is not consistent and might cause troubles when coding, like I'm experiencing.
> 
> > This issue prevents in practice the use of sympy with physical quantities for almost anything more than playing, I believe it's urgent to fix it.
> 
> Are you suggesting to leave `Mul(0, meter)` unevaluated? Maybe this is the quickest fix... but is it really the optimal fix?

I don't know which is the best way to fix it, I'm not expert enough of sympy to have an opinion on it, I just know what should be sympy behavior:

>>> import sympy
>>> from sympy.physics.units import *
>>> 0*meter
0*meter
>>> type(0*meter)
<class 'sympy.core.mul.Mul'>
>>> sympy.physics.units.util.convert_to(0*meter, cm)
0*centimeter

If the above behavior can be achieved through some quick and dirty fix I'll be glad to test it in an application I'm writing which makes deep use of physical quantities and just came to a stop due to such sympy issue.

**Comment by oscarbenjamin** [comment]
> Are you suggesting to leave `Mul(0, meter)` unevaluated? Maybe this is the quickest fix... but is it really the optimal fix?

Ideally `0` as a length should be distinct from `0` as a number.

The design of having units like `meter` as just args in a `Mul` seems like a mistake to me. It should be a separate type like `Quantity(0, meter)`. Either argument to `Quantity` could be a Mul but the idea that attaching units is just multiplication is conceptually flawed and all kinds of trouble in SymPy's unit system stem from this.

**Comment by Upabjojr** [comment]
> > Are you suggesting to leave `Mul(0, meter)` unevaluated? Maybe this is the quickest fix... but is it really the optimal fix?
> 
> Ideally `0` as a length should be distinct from `0` as a number.
> 
> The design of having units like `meter` as just args in a `Mul` seems like a mistake to me. It should be a separate type like `Quantity(0, meter)`. Either argument to `Quantity` could be a Mul but the idea that attaching units is just multiplication is conceptually flawed and all kinds of trouble in SymPy's unit system stem from this.

This is inherently connected with the difficulty of customizing SymPy's core expressions for ad-hoc use cases. SymPy has had this problem for a long time... there is no easy way to customize the way expressions behave in case modules external to SymPy's core (such as the units) are loaded.

**Comment by nopria** [comment]
> [...] the idea that attaching units is just multiplication is conceptually flawed and all kinds of trouble in SymPy's unit system stem from this.

Exactly. A physical quantity is not just a unit of measure multiplied with a number. The number is indeed just the "measure" of the unit, and cannot transform the physical quantity in something else, something dimensionless, just because it is zero. `0*meter` is not a multiplication, it's a measure of the unit 'meter' and remains a physical quantity regardless of the value of measure.

Is it so difficult to make `Mul` aware of that?

**Comment by nopria** [comment]
> For what it's worth, I think the second and third items on your list have other solutions:
> 
>     * A function to determine whether all values in a list have the same unit: a zero would be compatible with any unit, and thus would return "true" in unit comparisons with any quantity.

This is not simple as it seems. If I have to parse an input which may or may not have units (because I let the user employ a default unit by entering just the number), I have to way to understand if zero has been entered as a number which should be used with the default unit or has been entered with another unit preferred by the user.


> 
>     * If zeros are treated as compatible with any unit, this means that the unit conversion function would simply leave the zero as a zero.

Maybe I don't understand well your suggestion, but suppose you have a physical quantity and you need to separate magnitude from unit (I do it with `as_coeff_Mul()`, am I right?) in order to process (in some way) them separately; if the physical quantity derive from `0*meter` how can you process the unit if you don't know it? (because the multiplication with zero blew it away)

**Comment by Upabjojr** [comment]
What about this:
```python
class UnitZero(Quantity):
    _op_priority = 12.0  # necessary for SymPy's core

    def __init__(self, dimension):
        self._dimension = dimension

    def __mul__(self, other):
        return self

    ... similar for div add sub truediv pow rmul radd rdiv rtruediv ...
```
At this point `Quantity` can be modified as:
```python
class Quantity(...):
    _op_priority = 12.0

    def __mul__(self, other):
        if other == 0:
            return UnitZero(self. _get_dimension ... () )
        ... the rest ...

    ... other method overloading ...
```

**Comment by nopria** [comment]
After line [686 of Mul.py](https://github.com/sympy/sympy/blob/00d1e50fe5d014a1df7fcb37813d23f7061538d5/sympy/core/mul.py#L686) (`elif coeff.is_zero:`) there is a comment:

```
# we know for sure the result will be 0 except the multiplicand
# is infinity or a matrix
```

Doesn't it mean that there are exceptions to the rule `0*anything = 0` ? Can `Quantity` objects be added to such already existing exceptions?

**Comment by oscarbenjamin** [comment]
There are already too many special cases like this. We can't just hard-code everywhere like
```
if not matrix and not units and not ...
```
The kind system is intended as a way of positively distinguishing things like number vs matrix etc:
```python
In [1]: FiniteSet(1, 2).kind
Out[1]: SetKind(NumberKind)

In [2]: S(2).kind
Out[2]: NumberKind

In [3]: Matrix([1, 2]).kind
Out[3]: MatrixKind(NumberKind)
```
Ideally the check in Mul should be like:
```
if self.kind is NumberKind:
```
rather than trying to exhaustively check for things that are not numbers.

**Comment by nopria** [comment]
Why don't just set (internally to `Mul`) `evaluate=False` if one of the multiplicand is zero?

**Comment by oscarbenjamin** [comment]
> Why don't just set (internally to `Mul`) `evaluate=False` if one of the multiplicand is zero?

But then how would this evaluate:
```
In [1]: Mul(0, 2)
Out[1]: 0
```
Everything would break if that was changed.

**Comment by nopria** [comment]
> > Why don't just set (internally to `Mul`) `evaluate=False` if one of the multiplicand is zero?
> 
> But then how would this evaluate:
> 
> ```
> In [1]: Mul(0, 2)
> Out[1]: 0
> ```
> 
> Everything would break if that was changed.

Of course I didn't mean that, I forgot "and the other a Quanity".

**Comment by oscarbenjamin** [comment]
I already explained above:

> There are already too many special cases like this. We can't just hard-code everywhere like

**Comment by nopria** [comment]
> I already explained above:
> 
> > There are already too many special cases like this. We can't just hard-code everywhere like

I think that `0*quantity=0` is a serious flaw and should be fixed *whatever it takes*.

**Comment by asmeurer** [comment]
> Ideally the check in Mul should be like:
> 
> ```
> if self.kind is NumberKind:
> ```
> 
> rather than trying to exhaustively check for things that are not numbers.

What should the logic in Mul actually look like then? You mean that everything that's there currently only happens if the kind is NumberKind, and it's completely unevaluated otherwise?

**Comment by oscarbenjamin** [comment]
It should be something like:
```
if coeff.is_zero:
    if all(arg.kind is NumberKind for arg in args):
        if all(arg.is_finite for arg in args):
            return zero
```
This will not work though because many expressions that should be NumberKind are UndefinedKind so it would need to be:
```
if coeff.is_zero:
    if all(arg.kind in (NumberKind, UndefinedKind) for arg in args):
        if all(arg.is_finite for arg in args):
            return zero
```
Then Quantity should set its kind to QuantityKind (or something like that).

The point of `.kind` is supposed to be that we should have a way of partitioning expressions into mutually exclusive types. Then instead of
```
if x is not matrix and not quaternion and not quantity ...
    # presume that x is a number
```
we should be able to write:
```
if x is a number:
    # know that x is a number
```

**Comment by asmeurer** [comment]
Should we be using isinstance/class hierarchies for the kinds?

**Comment by oscarbenjamin** [comment]
> Should we be using isinstance/class hierarchies for the kinds?

I'm not sure exactly what you mean but distinct kinds are mutually exclusive and there is no concept of a "subkind" like there is for a subclass. Either the .kind is NumberKind or it is not. There can be no superkind or subkind of NumberKind, only other kinds that represent disjoint sets of objects.

**Comment by schymans** [comment]
> > Why don't just set (internally to `Mul`) `evaluate=False` if one of the multiplicand is zero?
> 
> But then how would this evaluate:
> 
> ```
> In [1]: Mul(0, 2)
> Out[1]: 0
> ```
> 
> Everything would break if that was changed.

Ehm, that was a misquote. @nopria said "if one of the multiplicand is zero **and the other a Quantity**" 
Why not make this simple exception for quantities only instead of trying to make multiplication only work for NumberKind and UndefindKind?
